[ML] Add api to update trained model deployment number_of_allocations

dimitris-athanasiou opened the pull request:
This commit adds a new API that users can use calling:

```
POST _ml/trained_models/{model_id}/deployment/_update
{
  "number_of_allocations": 4
}
```

This allows a user to update the number of allocations for a deployment that is `started`.

If the allocations are increased we rebalance and let the assignment planner find how to allocate the additional allocations.

If the allocations are decreased we cannot use the assignment planner. Instead, we implement the reduction in a new class `AllocationReducer` that tries to reduce the allocations so that:

  1. availability zone balance is maintained
  2. assignments that can be completely stopped are preferred to release memory

dimitris-athanasiou committed:
[ML] Add api to update trained model deployment number_of_allocations

This commit adds a new API that users can use calling:

```
POST _ml/trained_models/{model_id}/deployment/_update
{
  "number_of_allocations": 4
}
```

This allows a user to update the number of allocations for a deployment
that is `started`.

If the allocations are increased we rebalance and let the assignment
planner find how to allocate the additional allocations.

If the allocations are decreased we cannot use the assignment planner.
Instead, we implement the reduction in a new class `AllocationReducer`
that tries to reduce the allocations so that:

  1. availability zone balance is maintained
  2. assignments that can be completely stopped are preferred to release memory

elasticsearchmachine commented:
Hi @dimitris-athanasiou, I've created a changelog YAML for you.

dimitris-athanasiou committed:
Update docs/changelog/90728.yaml

lcawl commented on docs/reference/ml/trained-models/apis/update-trained-model-deployment.asciidoc:
```diff
@@ -0,0 +1,89 @@
+You can update a trained model deployment whose `assignment_state` is `started`.
+You can either increase or decrease the number of allocations of such a deployment.
+
+[[ml-update-datafeed-path-parms]]
```
```suggestion
[[ml-update-trained-model-deployments-path-parms]]
```

lcawl committed:
Update docs/reference/ml/trained-models/apis/update-trained-model-deployment.asciidoc

lcawl commented on docs/reference/ml/trained-models/apis/ml-trained-models-apis.asciidoc:
```diff
@@ -15,6 +15,7 @@ You can use the following APIs to perform model management operations:
 * <<infer-trained-model>>
 * <<start-trained-model-deployment>>
 * <<stop-trained-model-deployment>>
+* <<update-trained-model-deployment>>
```
```suggestion
* <<ml-update-trained-model-deployment>>
```

lcawl committed:
Update docs/reference/ml/trained-models/apis/ml-trained-models-apis.asciidoc

lcawl commented on rest-api-spec/src/main/resources/rest-api-spec/api/ml.update_trained_model_deployment.json:
```diff
@@ -0,0 +1,34 @@
+      ]
+    },
+    "body":{
+      "description":"The datafeed update settings",
```
```suggestion
      "description":"The updated trained model deployment settings",
```

lcawl committed:
Update rest-api-spec/src/main/resources/rest-api-spec/api/ml.update_trained_model_deployment.json

lcawl commented on docs/reference/ml/trained-models/apis/update-trained-model-deployment.asciidoc:
```diff
@@ -0,0 +1,89 @@
+
+Updates certain properties of a trained model deployment.
+
+
```
```suggestion
beta::[]

```

lcawl committed:
Update docs/reference/ml/trained-models/apis/update-trained-model-deployment.asciidoc

lcawl commented on docs/reference/ml/trained-models/apis/update-trained-model-deployment.asciidoc:
```diff
@@ -0,0 +1,90 @@
+[[ml-update-trained-model-deployment-request-body]]
+== {api-request-body-title}
+
+The following property can be updated after the trained model deployment is started:
```
IMO this text is unnecessary, since it's already mentioned in the description:
```suggestion

```

lcawl reviewed (approved):
One suggestion, otherwise docs LGTM

dimitris-athanasiou committed:
Remove unnecessary sentence and ml- prefix from docs

dimitris-athanasiou commented:
@lcawl Thank you the fixes! I removed the `ml-` prefix from the labels so that it is consistent with the other docs for trained models. I also removed that redundant sentence.

dimitris-athanasiou commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'main' into update-trained-model-deployment-api

dimitris-athanasiou committed:
Fix test

dimitris-athanasiou commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'main' into update-trained-model-deployment-api

dimitris-athanasiou committed:
Fix tests

droberts195 commented on x-pack/plugin/ml/qa/native-multi-node-tests/src/javaRestTest/java/org/elasticsearch/xpack/ml/integration/PyTorchModelIT.java:
```diff
@@ -877,14 +877,94 @@ public void testStartDeployment_GivenNoProcessorsLeft_AndLazyStartEnabled() thro
+    }
+
+    public void testUpdateDeployment_GivenAllocationsAreIncreasedOverResources_AndScalingIsPossible() throws Exception {
+        Request loggingSettings = new Request("PUT", "_cluster/settings");
+        loggingSettings.setJsonEntity("""
+            {"persistent" : {
+                    "xpack.ml.max_lazy_ml_nodes": 5
+                }}""");
+        client().performRequest(loggingSettings);
```
This isn't a logging setting, and it doesn't get reset after the test.

I think there's a danger that this will cause some other test to unexpectedly fail depending on the order that tests run.

dimitris-athanasiou replied:
There is a setup method in that test suite that unsets this setting as we also set it in other tests.

droberts195 commented on x-pack/plugin/src/yamlRestTest/resources/rest-api-spec/test/ml/update_trained_model_deployment.yml:
```diff
@@ -0,0 +1,10 @@
+---
+"Test with unknown model id":
+  - do:
+      catch: missing
```
Since the only YAML test throws an exception the clients flight recorder won't be able to record the success response format.

Would it be possible to incorporate an update into the middle of one of the tests in `3rd_party_deployment.yml` that starts a deployment?

dimitris-athanasiou replied:
Definitely. I'll add one.

droberts195 replied:
I couldn't see where this was added.

dimitris-athanasiou replied:
Sorry, I forgot to pick that commit from another branch where I added it. It should be present now in `3rd_party_deployment.yml`.

droberts195 commented on x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/assignment/TrainedModelAssignmentClusterService.java:
```diff
@@ -555,6 +556,148 @@ private boolean isScalingPossible(List<DiscoveryNode> nodes) {
+        ActionListener<ClusterState> listener
+    ) {
+        TrainedModelAssignment.Builder updatedAssignment = numberOfAllocations < assignment.totalTargetAllocations()
+            ? new AllocationReducer(assignment, nodeAvailabilityZoneMapper.getMlNodesByAvailabilityZone()).reduceTo(numberOfAllocations)
```
There's a potential discrepancy in cluster states used for different parts of the code here.  `nodeAvailabilityZoneMapper` keeps itself up to date by listening for cluster state updates.  But this method runs as part of a cluster state update that is supposed to work off of an exact version of the cluster state.  Maybe `NodeAvailabilityZoneMapper` should have a method that can calculate and return ML nodes by availability zone from a specific `ClusterState` that's passed to it?  Or maybe there's a better way to solve this problem.

dimitris-athanasiou replied:
I have raised https://github.com/elastic/elasticsearch/pull/90770 to address this.

dimitris-athanasiou replied:
I have now rebased this PR and am using the new method where we pass the specific cluster state

dimitris-athanasiou committed:
Add yaml test that successfully updates a deployment

davidkyle commented on x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/assignment/planning/AllocationReducer.java:
```diff
@@ -0,0 +1,137 @@
+            }
+
+            Map.Entry<String, Integer> smallestAssignmentInLargestZone = largestZoneAssignmentsInAscendingOrder.get(0);
+            if (smallestAssignmentInLargestZone.getValue() == 1
+                || (smallestAssignmentInLargestZone.getValue() <= allocationsToRemove
+                    && (minAllocationsInOtherZones == 0
+                        || ((largestZoneAllocations - smallestAssignmentInLargestZone.getValue()) >= minAllocationsInOtherZones + 1)))) {
```
```suggestion
            boolean noAllocationsInOtherZones = allocationsPerZoneInAscendingOrder.size() == 1;
            if (smallestAssignmentInLargestZone.getValue() == 1
                || (smallestAssignmentInLargestZone.getValue() <= allocationsToRemove
                    && (noAllocationsInOtherZones
                        || ((largestZoneAllocations - smallestAssignmentInLargestZone.getValue()) > minAllocationsInOtherZones)))) {
```

I found this quite hard to follow could you add a comment please

dimitris-athanasiou replied:
I have refactored the check in a method with commentary. That should make it simpler to understand.

davidkyle reviewed (approved):
LGTM

dimitris-athanasiou committed:
Update x-pack/plugin/ml/qa/native-multi-node-tests/src/javaRestTest/java/org/elasticsearch/xpack/ml/integration/PyTorchModelIT.java

Co-authored-by: David Kyle <david.kyle@elastic.co>

dimitris-athanasiou committed:
Update x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/assignment/planning/AllocationReducer.java

Co-authored-by: David Kyle <david.kyle@elastic.co>

dimitris-athanasiou committed:
Refactor complicated if condition

dimitris-athanasiou commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'main' into update-trained-model-deployment-api

dimitris-athanasiou committed:
Build ML nodes by zone using consistent cluster state

dimitris-athanasiou committed:
Make YAML update test update to 1

droberts195 reviewed (approved):
LGTM
